### PR TITLE
 [Feature][BugFix][OP] Enhance Deterministic Inference Mode with Kernel-level Fixes and Batch-invariant BMM

### PR DESCRIPTION
**Pull request by gongweibao**
## Motivation

LLM inference in FastDeploy can produce non-deterministic outputs across runs due to floating-point accumulation order differences in attention kernels, race conditions in penalty kernels, RNG offset drift in sampling, and missing batch-invariant BMM coverage. This PR comprehensively addresses these issues to make `FD_DETERMINISTIC_MODE=1` produce bit-exact results across identical inputs and runs.

Key goals:
1. Fix race conditions and non-determinism at the CUDA kernel level.
2. Extend batch-invariant operator coverage to include BMM operations.
3. Ensure Flash Attention uses deterministic accumulation order.
4. Stabilize sampling RNG behavior across generate() calls.
5. Add comprehensive unit tests and debugging infrastructure for determinism verification.

### CUDA Kernel Fixes

- **Append Attention (c16 + decoder)**: When `FD_DETERMINISTIC_MODE=1`, force the nosplit kernel path (`force_no_partition`) to guarantee consistent floating-point accumulation order regardless of sequence length. Fixed the c16 warp1_4 kernel to correctly handle the nosplit path when `num_chunks_this_seq > 1` by checking `partition_kv` template flag instead of only chunk count, preventing nullptr writes to `tmp_workspace`.
- **Token Penalty Kernel (`token_penalty_multi_scores.cu`)**: Fixed a race condition between `atomicCAS` (prompt marking) and `atomicMax`+`atomicAdd` (generated counting) that could cause non-deterministic repeat penalty values. Split the loop into two passes with a `__syncthreads()` barrier.
- **Beam Search Softmax (`beam_search_softmax.cu`)**: Added missing `cudaStreamSynchronize` after `cudaMemcpyAsync` (D2H) to ensure host-side values (`beam_width_scalar`, `beam_group_num_scalar`) are valid before use.

### Flash Attention

- **`flash_attn_backend.py`**: For Flash Attention V3, force `num_splits=1` to prevent non-deterministic KV split across SMs, ensuring a fixed accumulation order.

### Batch-invariant Operators

- **New `bmm_persistent` kernel**: Added a Triton-based persistent BMM kernel (`bmm_kernel_persistent`) with deterministic tile traversal order (batch-major ordering). Supports bf16/fp16/fp32 with per-dtype tuned block configs.
- **`bmm_batch_invariant`**: Drop-in replacement for `paddle._C_ops.bmm`, registered in `enable_batch_invariant_mode()`.

### Sampling Determinism

- **`top_k_top_p_sampling.py`**: In deterministic mode, reset the CUDA generator to a fixed seed before every sampling call to eliminate global RNG offset drift across `generate()` invocations, making `topp_seed` the sole source of per-request randomness.

### Deterministic Logging & Debugging

- **`deterministic_logger.py`**: Refactored to provide file-based MD5 hash collection (`/tmp/fd_det_logits_md5.jsonl`) and lightweight fingerprints (`/tmp/fd_det_logits.jsonl`). Added `_record_logits_diagnostic` helper for recording both bit-exact MD5 and human-readable summaries in a single GPU sync.
- **`deterministic_probe.py`** (new): Per-layer zero-sync probe for determinism debugging, gated by `FD_DETERMINISTIC_PROBE_LAYERS` env var.
- **`sampler.py`**: Integrated deterministic logging hooks for raw logits and post-penalty logits when `FD_DETERMINISTIC_LOG_MODE=1`.
- **`qwen2.py`**: Added optional per-layer hidden state recording via the deterministic probe.
- **New env var `FD_DETERMINISTIC_PROBE_LAYERS`**: Enables per-layer probing in model forward for isolating non-determinism sources.

### Worker Integration

- Worker process and GPU model runner updated to propagate deterministic mode flags and integrate with diagnostic infrastructure.

### Tests

- **`test_c16_warp1_4_determinism.py`**: Verifies the c16 warp1_4 nosplit kernel path produces bit-exact results.
- **`test_determinism_long.py`**: End-to-end determinism verification for long sequence generation.
- **`test_flash_attn_determinism.py`**: Validates Flash Attention determinism with `num_splits=1`.
- **`test_penalty_kernel_determinism.py`**: Verifies the penalty kernel race condition fix.
- **`test_sampling_determinism.py`**: Tests sampling RNG reset behavior across multiple calls.
- **`test_batch_invariance_op_bmm.py`**: Unit tests for the new batch-invariant BMM kernel.
- **`test_determinism_offline.py`**: Simplified offline determinism test.

## Usage or Command

Enable deterministic inference mode by setting the environment variable:

```bash
export FD_DETERMINISTIC_MODE=1
```

For debugging non-determinism with additional diagnostics:

```bash
export FD_DETERMINISTIC_MODE=1
export FD_DETERMINISTIC_LOG_MODE=1       # Enable MD5 hash logging
export FD_DETERMINISTIC_PROBE_LAYERS=1   # Enable per-layer probing
export FD_DETERMINISTIC_DEBUG=1          # Enable CUDA-side debug prints
```

Running determinism tests:

```bash
# Run all determinism tests
python -m pytest tests/deterministic/ -v

# Run specific tests
python -m pytest tests/deterministic/test_penalty_kernel_determinism.py -v
python -m pytest tests/deterministic/test_flash_attn_determinism.py -v
python -m pytest tests/deterministic/test_sampling_determinism.py -v
python -m pytest tests/batch_invariant/test_batch_invariance_op_bmm.py -v
```

## Accuracy Tests

- Determinism tests verify bit-exact (MD5 hash match) reproducibility across multiple runs.
- The penalty kernel fix resolves a correctness issue where `atomicCAS` and `atomicMax+atomicAdd` raced on the same slot, potentially producing incorrect repeat penalty scores.
- Flash Attention V3 `num_splits=1` produces mathematically identical results to the default split behavior for any given input, but with a fixed accumulation order.
- The batch-invariant BMM kernel uses persistent tiling with deterministic tile traversal, producing identical results regardless of batch composition.

## Checklist

- [x] Add at least a tag in the PR title.
  - Tags: `[Feature]`, `[BugFix]`, `[OP]`
- [x] Format your code, run `pre-commit` before commit.
- [x] Add unit tests. 7 new/updated test files covering all major changes.
- [x] Provide accuracy results. Determinism tests verify bit-exact MD5 hash reproducibility.
- [x] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

File: fastdeploy/model_executor/layers/sample/sampler.py [hunk 1]
```diff
@@ -498,6 +499,12 @@ def forward_cuda(
         p_done_idxs: List[int] = [],
     ) -> SamplerOutput:
         """ """
+        # Record raw logits fingerprint for determinism debugging
+        from fastdeploy import envs
```
Comment (gongshaotian) [inline fastdeploy/model_executor/layers/sample/sampler.py]: import 插到最前面

File: fastdeploy/model_executor/layers/attention/flash_attn_backend.py [hunk 2]
```diff
@@ -167,6 +177,7 @@ def flash_attn_func(
                 max_seqlen_q=max_seqlen_q,
                 max_seqlen_k=max_seqlen_k,
                 causal=causal,
+                num_splits=1,
```
Comment (gongshaotian) [inline fastdeploy/model_executor/layers/attention/flash_attn_backend.py]: 缺了个 _is_deterministic_mode 判断？，直接写死会影响非 deterministic吧

File: fastdeploy/model_executor/layers/batch_invariant_ops/batch_invariant_ops.py [hunk 3]
```diff
@@ -552,15 +741,47 @@ def enable_batch_invariant_mode():
...
+
+        paddle._C_ops.linear = _tracked_linear
+
+    if _original_ops["linear_v2"] is not None:
+        _orig_linear_v2 = _original_ops["linear_v2"]
+
+        def _tracked_linear_v2(*args, **kwargs):
+            return _orig_linear_v2(*args, **kwargs)
+
+        paddle._C_ops.linear_v2 = _tracked_linear_v2
```
Comment (gongshaotian) [inline fastdeploy/model_executor/layers/batch_invariant_ops/batch_invariant_ops.py]: 没理解这些  tracked func的作用

File: fastdeploy/model_executor/layers/sample/ops/top_k_top_p_sampling.py [hunk 4]
```diff
@@ -62,6 +69,14 @@ def top_k_top_p_sampling(
     """
     top_p_class = envs.FD_SAMPLING_CLASS.lower()
 
+    # In deterministic mode, reset CUDA generator offset before sampling.
+    # paddle.tensor.top_p_sampling uses the global GPU generator offset even
+    # when topp_seed is provided, causing RNG drift across generate() calls.
+    # Resetting to a fixed seed ensures the offset is always 0, making
+    # topp_seed the sole source of per-request randomness.
+    if envs.FD_DETERMINISTIC_MODE:
+        _reset_cuda_generator_for_determinism()
```
Comment (gongshaotian) [inline fastdeploy/model_executor/layers/sample/ops/top_k_top_p_sampling.py:78]: 为什么不放在初始化阶段做，每次调用都要reset吗

File: custom_ops/gpu_ops/token_penalty_multi_scores.cu
Comment (gongshaotian) [inline custom_ops/gpu_ops/token_penalty_multi_scores.cu:1]: 没看懂这个修改的目的是什么。我看到的是引入问题又自己解决了

File: custom_ops/gpu_ops/beam_search_softmax.cu
Comment (gongshaotian) [inline custom_ops/gpu_ops/beam_search_softmax.cu:1]: 这个算子对 Deterministic 是必需的吗？

File: custom_ops/gpu_ops/append_attn/multiquery_attention_c4_impl.cuh
Comment (gongshaotian) [inline custom_ops/gpu_ops/append_attn/multiquery_attention_c4_impl.cuh:1]: c8 c4的算子为啥没改 num_chunks？那 Deterministic 的单测是怎么通过的，单测是有效的吗？

File: custom_ops/gpu_ops/append_attn/multiquery_attention_c16_impl.cuh [hunk 5]
```diff
@@ -932,9 +938,24 @@ void MultiQueryAppendAttention(
 
     uint32_t chunk_size = static_cast<uint32_t>(encoder_max_partition_size);
     const int num_chunks = div_up(max_dec_len, chunk_size);
+    // Deterministic mode: force use nosplit kernel to ensure consistent
+    // floating-point accumulation order across all sequence lengths
+    const bool force_no_partition = getBoolEnv("FD_DETERMINISTIC_MODE");
+
+    // Debug log for determinism verification
+    if (getBoolEnv("FD_DETERMINISTIC_DEBUG")) {
```
Comment (gongshaotian) [inline custom_ops/gpu_ops/append_attn/multiquery_attention_c16_impl.cuh]: 这不是 get 过了，为啥又get

**Review by gongshaotian — CHANGES_REQUESTED** [review]

File: custom_ops/gpu_ops/beam_search_softmax.cu
Comment (gongshaotian) [reply custom_ops/gpu_ops/beam_search_softmax.cu:1]: beam search softamx 这个算子是不是已经弃用了@freeliuzc

File: fastdeploy/model_executor/layers/batch_invariant_ops/batch_invariant_ops.py (hunk 3 above)
Comment (gongweibao) [reply fastdeploy/model_executor/layers/batch_invariant_ops/batch_invariant_ops.py]: 这是一部分调试代码，用来记录算子的，检查是否有算子没有被替换掉。我去掉这些调试代码，放到另外的地方去。感谢！

File: fastdeploy/model_executor/layers/sample/ops/top_k_top_p_sampling.py (hunk 4 above)
Comment (gongweibao) [reply fastdeploy/model_executor/layers/sample/ops/top_k_top_p_sampling.py:78]: 对。paddle.tensor.top_p_sampling 这个函数，即使传了 topp_seed（每个请求自己的随机种子），它底层还是会读取全局 GPU 随机数生成器的 offset。这就导致一个问题——每调一次 generate()，这个全局 offset 就会往前走一点，造成随机数漂移。同样的输入，第一次调和第二次调，结果可能不一样。

解决办法：在每次采样前，把全局随机种子重置为一个固定值，这样 offset 就回到 0。这样一来，真正决定随机性的就只有 topp_seed（每个请求自带的种子），而不会受全局状态的干扰，从而保证相同输入 → 相同输出。

File: custom_ops/gpu_ops/token_penalty_multi_scores.cu
Comment (gongweibao) [reply custom_ops/gpu_ops/token_penalty_multi_scores.cu:1]: 之前的写法有竞态，，我看是[Sunny-bot1](https://github.com/gongweibao/FastDeploy/commits?author=Sunny-bot1) 改的。不是我改的。

File: custom_ops/gpu_ops/beam_search_softmax.cu
Comment (gongweibao) [reply custom_ops/gpu_ops/beam_search_softmax.cu:1]: 不是必须。是看到了一个潜在的风险，顺手就改了。

File: custom_ops/gpu_ops/token_penalty_multi_scores.cu
Comment (gongweibao) [reply custom_ops/gpu_ops/token_penalty_multi_scores.cu:1]: atomicMax 和 atomicAdd 虽然各自是原子的，但它们合在一起不是原子的。 有竞态条件。

File: custom_ops/gpu_ops/append_attn/multiquery_attention_c16_impl.cuh (hunk 5 above)
Comment (gongweibao) [reply custom_ops/gpu_ops/append_attn/multiquery_attention_c16_impl.cuh]: 你说的有道理。👍

File: fastdeploy/model_executor/layers/attention/flash_attn_backend.py (hunk 2 above)
Comment (gongweibao) [reply fastdeploy/model_executor/layers/attention/flash_attn_backend.py]: 这个地方其实不用改。FA3 本身就是num_splits=1。不过还是加上更直观一些。

File: custom_ops/gpu_ops/append_attn/multiquery_attention_c4_impl.cuh
Comment (gongweibao) [reply custom_ops/gpu_ops/append_attn/multiquery_attention_c4_impl.cuh:1]: 改了。原始
```
if (num_chunks <= 1) {
```
改成：
```
if (num_chunks <= 1 || force_no_partition)
```
c16 是因为之前的判断是一个死代码，没有打开过，所以需要改。c4等是因为`num_chunks <= 1`之前就有，所以记上或的条件就好了。

Comment (gongweibao) [reply custom_ops/gpu_ops/append_attn/multiquery_attention_c4_impl.cuh:1]: 很赞。这个确实没有加上，因为只有一行并且只是把之前的代码打开，忽略了。现在加上。

File: fastdeploy/model_executor/layers/sample/sampler.py (hunk 1 above)
Comment (gongweibao) [reply fastdeploy/model_executor/layers/sample/sampler.py]: Done

**Review by gongshaotian — APPROVED** [review]
